samza-test/src/main/java/org/apache/samza/test/framework/stream/CollectionStream.java

```diff
@@ -0,0 +1,189 @@
+ *        and value represents the message of the same
+ */
+public class CollectionStream<T> {
+
+  private String streamName;
+  private String systemName;
+  private CollectionStreamSystem system;
+  private Map<Integer, Iterable<T>> partitions;
+  private Map<String, String> streamConfig;
+  private static final String STREAM_TO_SYSTEM = "streams.%s.samza.system";
+
+  /**
+   * Constructs a new CollectionStream with multiple empty partitions from specified components.
+   * @param systemName represents name of the system stream is associated with
+   * @param streamName represents name of the stream
+   * @param partitionCount represents number of partitions, each of these partitions will be empty
+   */
+  private CollectionStream(String systemName, String streamName, Integer partitionCount) {
+    this.streamName = streamName;
+    this.streamConfig = new HashMap<>();
+    this.partitions = new HashMap<>();
+    this.systemName = systemName;
+    streamConfig.put(String.format(STREAM_TO_SYSTEM, this.streamName), systemName);
+    for (int i = 0; i < partitionCount; i++) {
+      partitions.put(i, new ArrayList<>());
+    }
+  }
+
+  /**
+   * Constructs a new CollectionStream with single partition from specified components.
+   * @param systemName represents name of the system stream is associated with
+   * @param streamName represents name of the stream
+   * @param partition represents the messages that the stream will be intialized with, default partitionId for the
+   *                  this single partition stream is 0
+   */
+  private CollectionStream(String systemName, String streamName, Iterable<T> partition) {
+    this.streamName = streamName;
+    this.streamConfig = new HashMap<>();
+    partitions = new HashMap<>();
+    this.systemName = systemName;
+    streamConfig.put(String.format(STREAM_TO_SYSTEM, this.streamName), systemName);
+    partitions.put(0, partition);
+  }
+
+  /**
+   * Constructs a new CollectionStream with multiple partitions from specified components.
+   * @param systemName represents name of the system stream is associated with
+   * @param streamName represents name of the stream
+   * @param partitions represents the partition state, key of the map represents partitionId and value represents
+   *                   the messages that partition will be initialized with
+   */
+  private CollectionStream(String systemName, String streamName, Map<Integer, ? extends Iterable<T>> partitions) {
+    this.streamName = streamName;
+    this.streamConfig = new HashMap<>();
+    this.partitions = new HashMap<>();
+    this.systemName = systemName;
+    streamConfig.put(String.format(STREAM_TO_SYSTEM, this.streamName), systemName);
+    partitions.entrySet().forEach(entry -> {
+        this.partitions.put(entry.getKey(), entry.getValue());
+      });
+  }
+
+  public Map<Integer, Iterable<T>> getInitPartitions() {
```

Comment:
Is this only used for input streams? It seems a bit inconsistent if it is accessible for both input and output streams, but it should not be used for output streams. In my opinion, the term "init" here can still be misinterpreted as "the first set of data produced to the stream", so getInitPartitions might still be misleading.
It seems to me like a stream instance can own all of the partition data associated with it, not just the partition data it was constructed with. What if you have a general getPartitions and addData, so it can work more consistently across both input and output? Then, instead of consumeStream returning the partition data, you could add it into the output stream object, and then read it out of the stream object.

Reply:
This method is used internally and is only used to initialize the input streams, will add javadoc to describe more. 

The problem is a CollectionStreamSystem can have multiple streams.
TestRunner constructs the CollectionStreamSystem and initializes all the CollectionStream associated with it. Since TestRunner maintains a Map of CollectionStreamSystem here  consumeStream can only belong to TestRunner.

CollectionStream in general just represents a description of stream and system it belongs to. Hence a general getPartitions is not possible unless a CollectionStream has instance of System it belongs to.

Reply:
If CollectionStream in general is the just description of the stream/system, then does "initPartitions" belong to CollectionStream?
I think the root of my concern is that you have a field which should only be used in certain use cases of the class (i.e. during input initialization). Ideally, you could use syntax (as opposed to convention) to force the usage only where it should be used. This isn't possible in all cases, but it feels like you could do that here.